### PR TITLE
Fix paddle.dist & paddle.nn.functional.normalize

### PR DESCRIPTION
**Pull request by fxyfxy777**
### PR Category

Operator Mechanism

### PR Types

Bug fixes

### 修复 `paddle.dist` 与 `paddle.nn.functional.normalize` 
两个算子共用一个kernel，所以放在一起分析
1. **前向计算报错（精度误差）**：
   - Paddle 前向输出为 `inf`，而 Torch 输出正常数值；
   - 原因：在 FP16 模式下，`reduce_sum(pow(...))` 的中间计算过程中发生溢出；
  - 解决方案：将中间计算过程提升到 float32，最后再 cast 回 float16 输出，避免溢出。

2. **反向计算报错一般情况**：
   - 修改前向后，反向结果仍不一致，其中部分case出现以下情况
   - 现象：Paddle 可正常求导，Torch 输出为 `inf` 或 `nan`；
   - 提高精度（如使用 float32）后，Paddle 与 Torch 输出一致；
   - 示例（float16 精度报错）：
     ```python
     [accuracy error] backward paddle.nn.functional.normalize(x=Tensor([4, 5, 6, 3579139], "float16"), p=4)
     Not equal to tolerance rtol=0.01, atol=0.01
     Greatest absolute difference: nan
     ```
   - 示例（float32 精度一致）：
     ```python
     [Pass] paddle.nn.functional.normalize(x=Tensor([4, 5, 6, 3579139], "float32"), p=4)
     ```
  - 解决方案：这部分case写入到torch_error_skip。

3. **反向计算报错特殊情况**
-现象：Paddle 和 Torch 都能反向计算出结果，但差距极大（最大超过 160）。
-原因：查看torch源码，发现两者反向计算公式基本一致，但对特殊值 0 的处理不同。
-	Paddle：使用 1 / (norm + eps)，导致较大数值。
-	Torch：使用 mask 将这些位置梯度直接置为 0。
-	优化：参考 Torch 的处理方式，对 norm == 0 的位置直接 mask 为 0，误差从上百缩小至 0.2 以内。
示例前后对比：

 - **修改前**：
    ```text
    输入: Tensor([2281701, 1], dtype="float32"), axis=1
    最大误差: 163.0
    ```

  - **修改后**：
    ```text
    输入: Tensor([2281701, 1], dtype="float32"), axis=1
    最大误差: 0.0625

4.**最终修改内容**
	•	在 float16 模式下，将中间计算提至 float32 精度，避免溢出。
	•	修复整数类型溢出问题，使用 int64_t 替代潜在溢出的 int32。
	•	对于 Torch 输出 inf/nan 的反向 case，跳过精度对齐测试。
	•	优化反向 norm == 0 情况的处理方式，使用 mask 替代除法以减少误差。
Pcard-92269

### CODE REVIEW

File: paddle/phi/kernels/gpu/dist_kernel.cu [hunk 1]
```diff
@@ -63,6 +63,18 @@ struct PowFunctor {
   Ty p_order_;
 };
 
+template <typename Tx,
+          typename Ty,
+          typename Tout>  // Tx is high precision, Tout is low/out precision
+struct PowFunctor_high_precision {
```
Comment (lshpku) [inline paddle/phi/kernels/gpu/dist_kernel.cu]: 命名规则不要混用，应该用驼峰命名法，PowFunctorHighPrecision

File: paddle/phi/kernels/gpu/p_norm_grad_kernel.cu [hunk 2]
```diff
@@ -59,12 +61,36 @@ struct PNormGradFunctor {
...
                   int size) {
+    auto x_mt = x->template cast<MT>();
+    auto y_mt = y->template cast<MT>();
+    auto dy_mt = dy->template cast<MT>();
+
+    auto norm_pow = y_mt.pow(-this->porder);
+    auto mask_norm_nonzero = (y_mt != static_cast<MT>(0)).template cast<MT>();
+
+    // ⚠️ 新增：当 porder < 0 且 x == 0 的位置，置 0
+    // 条件掩码：x == 0 && this->porder < 0
```
Comment (lshpku) [inline paddle/phi/kernels/gpu/p_norm_grad_kernel.cu]: 不要用中文注释

Comment (fxyfxy777) [reply paddle/phi/kernels/gpu/p_norm_grad_kernel.cu]: 收到，我改一下

File: paddle/phi/kernels/gpu/p_norm_grad_kernel.cu [hunk 3]
```diff
@@ -59,12 +61,37 @@ struct PNormGradFunctor {
...
-        (*x).abs().pow(this->porder) * (*x).sign() * dy->broadcast(dim) *
-        (*y + y->constant(eps)).pow(-this->porder).broadcast(dim);
+    auto x_mt = x->template cast<MT>();
+    auto y_mt = y->template cast<MT>();
+    auto dy_mt = dy->template cast<MT>();
+
+    auto norm_pow = y_mt.pow(-this->porder);
+    auto mask_norm_nonzero = (y_mt != static_cast<MT>(0)).template cast<MT>();
+
+    // ⚠️ New: Set to 0 where porder < 0 and x == 0
```
Comment (lshpku) [inline paddle/phi/kernels/gpu/p_norm_grad_kernel.cu]: 也不要用非ascii的符号，另外注释应该站在读代码的人的立场上写，这个New在这里就让人摸不着头脑，因为这里不存在Old和New的对比

Comment (fxyfxy777) [reply paddle/phi/kernels/gpu/p_norm_grad_kernel.cu]: 好的好的

**Review by lshpku — APPROVED** [review]

**Review by wanghuancoder — APPROVED** [review]
LGTM
